### PR TITLE
fix: parallel parquet can underflow when max_record_batch_rows < execution.batch_size

### LINKED ISSUE #9736: Parallel parquet writer can panic when max_row_group_size < execution.batch_size
### Describe the bug

An underflow panic can sometimes be triggered by setting max_row_group_size to less than execution.batch_size. 

```
thread 'tokio-runtime-worker' panicked at datafusion/core/src/datasource/file_format/parquet.rs:886:33:
attempt to subtract with overflow
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

### To Reproduce

I found this by trying a wide range of max_row_group_sizes

```rust
#[tokio::main]
async fn main() -> Result<()> {
    // create local execution context
    let ctx = SessionContext::new();

    let testdata = "benchmarks/data/tpch_sf10/lineitem";

    let filename = &format!("{testdata}/part-0.parquet");

    // define the query using the DataFrame trait
    let df = ctx
        .read_parquet(filename, ParquetReadOptions::default())
        .await?
        .limit(0, Some(200_000))?;

    println!("{}", df.clone().count().await?);

    
    for row_group_size in (1..8193).step_by(283).rev(){
        println!("row group size: {}", row_group_size);
        println!("Writing without parallelism!");
        let row_group_path = format!("/tmp/{}.parquet", row_group_size);
        let mut options = TableParquetOptions::default();
        options.global.max_row_group_size = row_group_size;
        options.global.allow_single_file_parallelism = false;
        df.clone().write_parquet(
                &row_group_path,
                DataFrameWriteOptions::new().with_single_file_output(true),
                Some(options),
        )
        .await
        .unwrap();

        println!("Writing with parallelism!");
        let row_group_path = format!("/tmp/para_{}.parquet", row_group_size);
        let mut options = TableParquetOptions::default();
        options.global.max_row_group_size = row_group_size;
        options.global.allow_single_file_parallelism = true;
        df.clone().write_parquet(
                &row_group_path,
                DataFrameWriteOptions::new().with_single_file_output(true),
                Some(options),
        )
        .await
        .unwrap();

    }
    
    Ok(())
}
```

### Expected behavior

No combination of max_row_group_size and execution.batch_size should lead to panic

### Additional context

Extremely tiny max_row_group sizes can cause a stack overflow error even if parallel_parquet writer is disabled. E.g. a max row group size of 1. We may want to raise a configuration validation error for absurdly small row group sizes. 

This happens even if release mode:

```
row group size: 1
Writing without parallelism!

thread 'tokio-runtime-worker' has overflowed its stack
fatal runtime error: stack overflow
Aborted (core dumped)
```

### PR DESCRIPTION
**Pull request by devinjdangelo**
## Which issue does this PR close?

Closes #9736

## Rationale for this change

See issue

## What changes are included in this PR?

Parallel parquet writer can now handle the case when max_record_batch_rows < execution.batch_size by iteratively splitting the record batch rather than assuming it only needs to be split once. 

## Are these changes tested?

Yes, added new test that would panic prior to this PR

## Are there any user-facing changes?

Just bugfix

### CODE REVIEW

File: datafusion/core/src/dataframe/parquet.rs [hunk 1]
```diff
@@ -166,4 +166,47 @@ mod tests {
...
+    #[tokio::test]
+    async fn write_parquet_with_small_rg_size() -> Result<()> {
+        let mut test_df = test_util::test_table().await?;
+        // make the test data larger so there are multiple batches
+        for _ in 0..7 {
+            test_df = test_df.clone().union(test_df)?;
+        }
+        let output_path = "file://local/test.parquet";
+
+        for rg_size in (1..7).step_by(5) {
```
Comment (alamb) [inline datafusion/core/src/dataframe/parquet.rs:179]: My reading of the [docs](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.step_by) and my [playground experiments](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2ff81b06e8447214338607095c726d7c) suggests this is the same as `[1, 6]` -- is that the intent? Or did you mean `1, 5, 10, 15, 20, 25, 30, 35`?

File: datafusion/core/src/dataframe/parquet.rs [hunk 2]
```diff
@@ -166,4 +166,47 @@ mod tests {
...
+            options.global.allow_single_file_parallelism = true;
+            df.write_parquet(
+                output_path,
+                DataFrameWriteOptions::new().with_single_file_output(true),
+                Some(options),
+            )
+            .await?;
+
+            // Check that file actually used the correct rg size
+            let file = std::fs::File::open(tmp_dir.into_path().join("test.parquet"))?;
```
Comment (alamb) [inline datafusion/core/src/dataframe/parquet.rs:197]: Calling `into_path` here I think means the file won't be cleaned up 

I think calling `path()` would ensure the file is cleaned up

File: datafusion/core/src/dataframe/parquet.rs [hunk 3]
```diff
@@ -166,4 +166,47 @@ mod tests {
 
         Ok(())
     }
+
+    #[tokio::test]
+    async fn write_parquet_with_small_rg_size() -> Result<()> {
+        let mut test_df = test_util::test_table().await?;
+        // make the test data larger so there are multiple batches
+        for _ in 0..7 {
+            test_df = test_df.clone().union(test_df)?;
```
Comment (alamb) [inline datafusion/core/src/dataframe/parquet.rs:175]: When I ran this test it takes more than 22 seconds on my laptop. I wonder if we really need to generate so much data -- maybe we can try slicing up the batch (or else maybe use larger rg_sizes)

```shell
$ cargo test --lib -p datafusion -- write_parquet_with_small_rg_size
...
    Finished test [unoptimized + debuginfo] target(s) in 0.16s
     Running unittests src/lib.rs (target/debug/deps/datafusion-4cbfc61ad6017be4)

running 1 test
test dataframe::parquet::tests::write_parquet_with_small_rg_size ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 651 filtered out; finished in 22.31s

```

**Review by alamb — COMMENTED** [review]
Thank you @devinjdangelo  -- this code looks good to me. I think there are some small issues with the test to fix, but otherwise I think this is good to go.

🙏 

I ran the test without the changes in this PR and it fails like
```shell
b.rs (target/debug/deps/datafusion-4cbfc61ad6017be4)

attempt to subtract with overflow
thread 'dataframe::parquet::tests::write_parquet_with_small_rg_size' panicked at datafusion/core/src/datasource/file_format/parquet.rs:885:33:
attempt to subtract with overflow
stack backtrace:
```

File: datafusion/core/src/dataframe/parquet.rs (hunk 3 above)
Comment (devinjdangelo) [reply datafusion/core/src/dataframe/parquet.rs:175]: We should be able to trigger the issue with less data by lowering `execution.batch_size` to something small like 10 rows.

File: datafusion/core/src/dataframe/parquet.rs (hunk 1 above)
Comment (devinjdangelo) [reply datafusion/core/src/dataframe/parquet.rs:179]: Yes, [1, 6] is all I meant and would be clearer... I originally wanted to loop over more rg sizes but the test was slow. If we streamline the test, we could actually range over more values here.

Comment (devinjdangelo) [reply datafusion/core/src/dataframe/parquet.rs]: This now loops over 0..10 with datafusion.execution.batch_size set to 10.

File: datafusion/core/src/dataframe/parquet.rs (hunk 3 above)
Comment (devinjdangelo) [reply datafusion/core/src/dataframe/parquet.rs]: The test with a batch size of 10 still panics on main and passes in this PR, but runs in 0.41 seconds.

File: datafusion/core/src/dataframe/parquet.rs (hunk 2 above)
Comment (devinjdangelo) [reply datafusion/core/src/dataframe/parquet.rs]: I fixed this in the new test and a preexisting one.

File: datafusion/core/src/dataframe/parquet.rs [hunk 4]
```diff
@@ -150,7 +152,7 @@ mod tests {
             .await?;
 
             // Check that file actually used the specified compression
-            let file = std::fs::File::open(tmp_dir.into_path().join("test.parquet"))?;
+            let file = std::fs::File::open(tmp_dir.path().join("test.parquet"))?;
```
Comment (alamb) [inline datafusion/core/src/dataframe/parquet.rs:155]: 👍

Comment (alamb) [reply datafusion/core/src/dataframe/parquet.rs:155]: thank you for the driveby cleanup

**Review by alamb — APPROVED** [review]
Thanks @devinjdangelo -- I also verified the test now runs quickly and still panics 

Thank you so much

**Pull request merged**